pkg/querier/tripperware/shard_by.go

```diff
@@ -0,0 +1,97 @@
+package tripperware
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/thanos-io/thanos/pkg/querysharding"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"github.com/weaveworks/common/httpgrpc"
+
+	cquerysharding "github.com/cortexproject/cortex/pkg/querysharding"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	util_log "github.com/cortexproject/cortex/pkg/util/log"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+)
+
+func ShardByMiddleware(logger log.Logger, limits Limits, merger Merger) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return shardBy{
+			next:   next,
+			limits: limits,
+			merger: merger,
+			logger: logger}
+	})
+}
+
+type shardBy struct {
+	next          Handler
+	limits        Limits
+	logger        log.Logger
+	merger        Merger
+	queryAnalyzer *querysharding.QueryAnalyzer
+}
+
+func (s shardBy) Do(ctx context.Context, r Request) (Response, error) {
+	tenantIDs, err := tenant.TenantIDs(ctx)
+
+	if err != nil {
+		return nil, httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+	}
+
+	numShards := validation.SmallestPositiveIntPerTenant(tenantIDs, s.limits.QueryVerticalShardSize)
+
+	if numShards <= 1 {
+		return s.next.Do(ctx, r)
+	}
+
+	logger := util_log.WithContext(ctx, s.logger)
+	analysis, err := s.queryAnalyzer.Analyze(r.GetQuery())
+	if err != nil {
+		level.Warn(logger).Log("msg", "error sharding query", "q", r.GetQuery(), "err", err)
+	}
+
+	if err != nil || !analysis.IsShardable() {
+		return s.next.Do(ctx, r)
+	}
+
+	reqs := s.shardQuery(logger, numShards, r, analysis)
+
+	reqResps, err := DoRequests(ctx, s.next, reqs, s.limits)
+	if err != nil {
+		return nil, err
+	}
+
+	resps := make([]Response, 0, len(reqResps))
+	for _, reqResp := range reqResps {
+		resps = append(resps, reqResp.Response)
+	}
+
+	response, err := s.merger.MergeResponse(resps...)
+	if err != nil {
+		return nil, err
+	}
+	return response, nil
```

Comment:
We could simplify line 72 ~ 76 to `return s.merger.MergeResponse(resps...)`

Reply:
done